torch/nn/utils/parametrizations.py

```diff
@@ -0,0 +1,180 @@
+from .. import functional as F
+
+from typing import Optional
+
+class SpectralNorm(Module):
+    def __init__(
+        self,
+        weight: torch.Tensor,
+        n_power_iterations: int = 1,
+        dim: int = 0,
+        eps: float = 1e-12
+    ) -> None:
+        super().__init__()
+        self.dim = dim
+        if n_power_iterations <= 0:
+            raise ValueError('Expected n_power_iterations to be positive, but '
+                             'got n_power_iterations={}'.format(n_power_iterations))
+        self.n_power_iterations = n_power_iterations
+        self.eps = eps
+        self.register_buffer('u', None)
+        self.register_buffer('v', None)
+
+        weight_mat = self._reshape_weight_to_matrix(weight)
+        self._update_vectors(weight_mat)
+
+    def _reshape_weight_to_matrix(self, weight: torch.Tensor) -> torch.Tensor:
+        weight_mat = weight
+        if self.dim != 0:
+            # permute dim to front
+            weight_mat = weight_mat.permute(self.dim,
+                                            *[d for d in range(weight_mat.dim()) if d != self.dim])
+        height = weight_mat.size(0)
+        return weight_mat.reshape(height, -1)
+
+    @torch.autograd.no_grad()
+    def _update_vectors(self, weight_mat: torch.Tensor) -> None:
+        # See original note at torch/nn/utils/spectral_norm.py
+        # NB: If `do_power_iteration` is set, the `u` and `v` vectors are
+        #     updated in power iteration **in-place**. This is very important
+        #     because in `DataParallel` forward, the vectors (being buffers) are
+        #     broadcast from the parallelized module to each module replica,
+        #     which is a new module object created on the fly. And each replica
+        #     runs its own spectral norm power iteration. So simply assigning
+        #     the updated vectors to the module this function runs on will cause
+        #     the update to be lost forever. And the next time the parallelized
+        #     module is replicated, the same randomly initialized vectors are
+        #     broadcast and used!
+        #
+        #     Therefore, to make the change propagate back, we rely on two
+        #     important behaviors (also enforced via tests):
+        #       1. `DataParallel` doesn't clone storage if the broadcast tensor
+        #          is already on correct device; and it makes sure that the
+        #          parallelized module is already on `device[0]`.
+        #       2. If the out tensor in `out=` kwarg has correct shape, it will
+        #          just fill in the values.
+        #     Therefore, since the same power iteration is performed on all
+        #     devices, simply updating the tensors in-place will make sure that
+        #     the module replica on `device[0]` will update the _u vector on the
+        #     parallized module (by shared storage).
+        #
+        #    However, after we update `u` and `v` in-place, we need to **clone**
```

Comment:
I'm not really buying this. They are modified inplace with an `out=` function which does not support autograd. So I'm really surer how that can happen.
But I can we can check that out later.

Reply:
Ho that whole function is in no_grad mode. So no gradients here.
And indeed these will be saved in the forward so the clone is good.

Reply:
Oh, now that's interesting. I wonder whether it's possible to approximate the derivative with respect to the largest singular values of a matrix without computing the whole SVD... I'll think about this.

Reply:
Nevermind. It turns out that differentiating this expression with `u` and `v` fixed actually gives the correct gradients for the expression `A / \norm{A}_2`